File: app/src/main/java/com/routesme/taxi_screen/java/Tracking/Class/LocationTrackingService.java
package com.routesme.taxi_screen.java.Tracking.Class;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.util.Log;

import com.routesme.taxi_screen.java.Class.Helper;

import java.net.URI;
import java.net.URISyntaxException;

import tech.gusavila92.websocketclient.WebSocketClient;

public class LocationTrackingService {

    private Context context;
    private SharedPreferences sharedPreferences;
    private WebSocketClient trackingWebSocket;
    private TrackingHandler trackingHandler;
    private boolean isHandlerTrackingRunning = false;
    private Handler handlerTracking;
    private Runnable runnableTracking;


    public LocationTrackingService(Context context) {
        this.context = context;
        sharedPreferences = context.getSharedPreferences("userData", Activity.MODE_PRIVATE);
        createWebSocket();
        trackingHandler = new TrackingHandler(context, trackingWebSocket);
        startTracking();
    }


    private void createWebSocket() {
        String tabletSerialNo = sharedPreferences.getString("tabletSerialNo", null);
        setTrackingWebSocketConfiguration(trackingWebSocketUri(),tabletSerialNo);
    }

    private URI trackingWebSocketUri() {
        URI trackingWebSocketUri = null;
        try {
            trackingWebSocketUri = new URI(Helper.getConfigValue(context, "trackingWebSocketUri"));
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
        return trackingWebSocketUri;
    }

    private void setTrackingWebSocketConfiguration(URI trackingWebSocketUri, final String tabletSerialNo) {
        trackingWebSocket = new WebSocketClient(trackingWebSocketUri) {
            @Override
            public void onOpen() {
                Log.i("trackingWebSocket:  ", "Opened");
                //Send deviceId to server (Device Identifier)
                sendMessageViaSocket("deviceId:" + tabletSerialNo);
                //Send offline locations to server if it exists....
                trackingHandler.sendOfflineTrackingToServer();
                //Start Tracking Timer after 500 melli seconds...
                handlerTracking.post(runnableTracking);
            }

            @Override
            public void onTextReceived(String message) {
                Log.i("trackingWebSocket:  ", "Received message:  " + message);
            }

            @Override
            public void onBinaryReceived(byte[] data) {
            }

            @Override
            public void onPingReceived(byte[] data) {
            }

            @Override
            public void onPongReceived(byte[] data) {
            }

            @Override
            public void onException(Exception e) {
                Log.i("trackingWebSocket:  ", "Exception Error:   " + e.getMessage());
                stopTrackingTimer();
            }

            @Override
            public void onCloseReceived() {
                Log.i("trackingWebSocket:  ", "Closed !");
                stopTrackingTimer();
            }
        };
        trackingWebSocket.setConnectTimeout(10000);
        trackingWebSocket.setReadTimeout(60000);
        trackingWebSocket.enableAutomaticReconnection(5000);
    }


    private void sendMessageViaSocket(String message) {
        trackingWebSocket.send(message);
        Log.i("trackingWebSocket:  ", "Send message:  " + message);
    }

    private void startTracking() {
        LocationFinder locationFinder = new LocationFinder(context, trackingHandler);
        if (locationFinder.canGetLocation()) {
            setupTrackingTimer();
            trackingWebSocket.connect();
        } else {
            locationFinder.showSettingsAlert();
        }
    }


    private void setupTrackingTimer() {
        runnableTracking = new Runnable() {
            @Override
            public void run() {
                isHandlerTrackingRunning = true;
                Log.i("trackingWebSocket:  ", "Tracking Timer running ...");
                trackingHandler.locationChecker();
                handlerTracking.postDelayed(runnableTracking, 5000);
            }
        };
        handlerTracking = new Handler();
    }

    private void stopTrackingTimer() {
        if (isHandlerTrackingRunning) {
            handlerTracking.removeCallbacks(runnableTracking);
            isHandlerTrackingRunning = false;
            Log.i("trackingWebSocket:  ", "Tracking Timer stop ...");
        }
    }

    public void stopLocationTrackingService(){
        if (trackingWebSocket != null) {
            trackingWebSocket.onCloseReceived();
        }
    }
}

File: app/src/main/java/com/routesme/taxi_screen/kotlin/LocationTrackingService/Class/LocationTrackingService.kt
package com.routesme.taxi_screen.kotlin.LocationTrackingService.Class

import android.Manifest
import android.app.*
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Binder
import android.os.Build
import android.os.Handler
import android.os.IBinder
import android.util.Log
import androidx.core.content.ContextCompat
import com.routesme.taxi_screen.kotlin.Class.App
import com.routesme.taxi_screen.kotlin.Class.Helper
import com.routesme.taxi_screen.kotlin.Class.SharedPreference
import com.routesme.taxiscreen.R
import tech.gusavila92.websocketclient.WebSocketClient
import java.io.IOException
import java.net.URI

class LocationTrackingService(): Service() {

    private lateinit var trackingWebSocket: WebSocketClient
    private lateinit var trackingDataLayer: TrackingDataLayer
    private lateinit var locationReceiver: LocationReceiver
    private var isHandlerTrackingRunning = false
    private var handlerTracking: Handler? = null
    private var runnableTracking: Runnable? = null
    private val permissions = arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.READ_PHONE_STATE)
    private var handlerCheckPermissions: Handler? = null
    private var runnableCheckPermissions: Runnable? = null
    private var permissionsHandlerRunning = false

    private val sharedPreferences = App.instance.getSharedPreferences(SharedPreference.device_data, Activity.MODE_PRIVATE)
    private lateinit var authorityUrl: URI
    private var vehicleId: Int = -999
    private var institutionId: Int = -999
    private var deviceId: Int = -999

    var isWebSocketAlive: Boolean = false
    companion object {
        @get:Synchronized
        var instance:LocationTrackingService = LocationTrackingService()

        class LocationServiceBinder : Binder() {
            val service: LocationTrackingService
                get() = instance
        }
    }

    private val webSocket = object : WebSocketClient(getTrackingUrl()) {
        override fun onOpen() {
            isWebSocketAlive = true
            Log.d("Tracking-Logic", "WebSocket-onOpen")
        }
        override fun onTextReceived(message: String) {
            Log.d("Tracking-Logic", "WebSocket-onTextReceived: $message")
        }
        override fun onBinaryReceived(data: ByteArray) {}
        override fun onPingReceived(data: ByteArray) {}
        override fun onPongReceived(data: ByteArray) {}
        override fun onException(e: Exception) {
            if (e is IOException) {
                isWebSocketAlive = false
            }
            Log.d("Tracking-Logic", "WebSocket-onException")
        }
        override fun onCloseReceived() {
            isWebSocketAlive = false
            Log.d("Tracking-Logic", "WebSocket-onCloseReceived")
        }
    }

    private fun setTrackingWebSocketConfiguration(): WebSocketClient {

        webSocket.setConnectTimeout(10000)
        webSocket.setReadTimeout(60000)
        webSocket.enableAutomaticReconnection(10000)
        return webSocket
    }

    private fun getTrackingUrl(): URI {
        return URI(Helper.getConfigValue("trackingWebSocketUrl", R.raw.config))
       /*
        try {
           authorityUrl = URI(Helper.getConfigValue("trackingWebSocketAuthorityUrl",R.raw.config))
        }
       catch (e: URISyntaxException) {
            e.printStackTrace()
       }

        val builder: Uri.Builder = Uri.Builder()
       builder.scheme("http")
               .authority(authorityUrl.toString())
               .appendPath("trackServiceHub")
               .appendQueryParameter("vehicleId", vehicleId.toString())
               .appendQueryParameter("institutionId", institutionId.toString())
               .appendQueryParameter("deviceId", deviceId.toString())
       return URI(builder.build().toString())
    */
    }

    fun checkPermissionsGranted(){
        if (hasPermissions(*permissions)){
            startTracking()
        }else{
            setupCheckPermissionsHandler()
            permissionsHandlerRunning = true
            handlerCheckPermissions?.postDelayed(runnableCheckPermissions,1000)

        }
    }
    private fun setupCheckPermissionsHandler() {
        Log.i("trackingWebSocket:","setupCheckPermissionsHandler")
        runnableCheckPermissions = Runnable {
            if (hasPermissions(*permissions) && permissionsHandlerRunning ) {permissionsHandlerRunning = false; handlerCheckPermissions?.removeCallbacks(runnableCheckPermissions); instance.startTracking() }
            Log.i("trackingWebSocket:","startCheckPermissionsHandler")
            handlerCheckPermissions?.postDelayed(runnableCheckPermissions, 1 * 60 * 1000)
        }
        handlerCheckPermissions = Handler()
    }
     private fun startTracking() {
         vehicleId = getVehicleId()
         institutionId = getInstitutionId()
         deviceId = getDeviceId()
        testingSetup()
             if (vehicleId >= 0 && institutionId >= 0 && deviceId >= 0) {
                 trackingWebSocket = setTrackingWebSocketConfiguration()
                 trackingDataLayer = TrackingDataLayer(this, trackingWebSocket)
                 locationReceiver = LocationReceiver(trackingDataLayer)
                 if (locationReceiver.setUpLocationListener()) {
                     trackingWebSocket.connect()
                     setupTrackingHandler()
                     handlerTracking?.post(runnableTracking)
                 }
             }else return
    }

    private fun testingSetup() {
        vehicleId = 16
        institutionId = 9
        deviceId = 3
    }

    private fun hasPermissions(vararg permissions: String): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            for (p in permissions) {
                if (ContextCompat.checkSelfPermission(App.instance, p) != PackageManager.PERMISSION_GRANTED) {
                    return false
                }
            }
        }
        return true
    }

    private fun setupTrackingHandler() {
        runnableTracking = Runnable {
            isHandlerTrackingRunning = true

            Log.i("trackingWebSocket:  ", "Tracking Timer running ...")
            if (isWebSocketAlive) trackingDataLayer.executeTrackingLogic()
            handlerTracking?.postDelayed(runnableTracking, 5000)
        }
        handlerTracking = Handler()
    }

    override fun onBind(intent: Intent): IBinder {
        Log.i("trackingWebSocket:","onBind")
        return LocationServiceBinder()
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        super.onStartCommand(intent, flags, startId)
        Log.i("trackingWebSocket:","onStartCommand")
        return START_STICKY
    }

    override fun onCreate() {
        super.onCreate()
        startForeground(12345678, getNotification())
        Log.i("trackingWebSocket:","onCreate")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.i("trackingWebSocket:","onDestroy")
    }

    private fun getNotification(): Notification {
        val channel = NotificationChannel("channel_01", "My Channel", NotificationManager.IMPORTANCE_DEFAULT)
        val notificationManager = getSystemService(NotificationManager::class.java)
        notificationManager.createNotificationChannel(channel)
        val builder = Notification.Builder(applicationContext, "channel_01").setAutoCancel(true)
        return builder.build()
    }

    private fun getVehicleId() = sharedPreferences.getInt(SharedPreference.vehicle_id, -999)
    private fun getInstitutionId() = sharedPreferences.getInt(SharedPreference.institution_id, -999)
    private fun getDeviceId() =  sharedPreferences.getInt(SharedPreference.device_id, -999)
}